### ISSUE TITLE
Randomly getting stuck, wrong tensor dimensions, OOM

### ISSUE THREAD

**Issue by hohteri** [ROCmSoftwarePlatform/tensorflow-upstream#1059]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: -
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-30-g34c3143ce3 2.2.0
- Python version: Python 3.6.9
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): -
- CUDA/cuDNN version: -
- GPU model and memory: 2x Radeon VII 16GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

We run a training set which takes about 25 minutes with both GPUs active. Sometimes training passes, sometimes it just gets stuck, sometimes it exits OOM with tensor size ridiculously large. Seems similar to https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/1026 and https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/325 

**Describe the expected behavior**

Training would pass every time.

**Standalone code to reproduce the issue**

This is not code to reproduce the issue but you to see how we run the model:

Input is 312x1 in 3 channels

strategy = tf.distribute.MirroredStrategy()
    with strategy.scope():
            model = tf.keras.Sequential([
                tf.keras.layers.Conv1D(filters=opts.feature1Count, kernel_size=opts.kernel1Size, strides=1, pa
dding='same', dilation_rate=1, activation='relu', use_bias=True, kernel_initializer='glorot_uniform', bias_ini
tializer='zeros', input_shape=(opts.inputDim,opts.channelCount)),
                tf.keras.layers.MaxPooling1D(pool_size=opts.pool1Size, strides=None, padding='valid'),
                tf.keras.layers.Conv1D(filters=opts.feature2Count, kernel_size=opts.kernel2Size, strides=1, padding='same', dilation_rate=1, activation='relu', use_bias=True, kernel_initializer='glorot_uniform', bias_initializer='zeros'),
                tf.keras.layers.MaxPooling1D(pool_size=opts.pool2Size, strides=None, padding='valid'),
                tf.keras.layers.Flatten(),
                tf.keras.layers.Dense(opts.denseParameters, activation='relu', use_bias=True),
                tf.keras.layers.Dropout(opts.dropRate),
                tf.keras.layers.Dense(opts.outputDim, activation='softmax')
            ])
model.compile( optimizer=tfa.optimizers.Lookahead( tf.keras.optimizers.Adam(learning_rate=opts
.learningRate, beta_1=0.9, beta_2=0.999, amsgrad=True), sync_period=6, slow_step_size=0.5 ),
                           loss=tf.keras.losses.SparseCategoricalCrossentropy(),
                           metrics=[tf.keras.metrics.SparseCategoricalAccuracy()])

reduce_lr_callback = tf.keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=opts.ReduceLROnPlateau, patience=5, min_lr=1e-6, verbose=1, mode='min', min_delta=1e-4, cooldown=5)

history = model.fit(x=training_generator, class_weight=classWeight, steps_per_epoch=stepsPerEpoch_TRAIN, validation_data=validationDataset, validation_steps=None, verbose=1, callbacks=callbackList, epochs=opts.epochCount, max_queue_size=opts.queueSize, use_multiprocessing=opts.multiProcessing, workers=opts.workers, shuffle=True, initial_epoch=initialEpoch)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

If we run with AMD_SERIALIZE_KERNEL=3, AMD_SERIALIZE_COPY=3, TF_NUM_INTEROP_THREADS=1 like @ekuznetsov139 suggests in #1026 we get similar results, with AMD_xxx training gets stuck before it starts and TF_NUM... it seems succeed every time so far (of course training time is like 30% slower.

Attached is with TF_CPP_MIN_VLOG_LEVEL=3 one OOM and one random stickiness, along with rocminfo. Please note that our CPU is ancient but supported.

[rocminfo.txt](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/files/4977422/rocminfo.txt)
[OOM.txt](https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/files/4977423/OOM.txt)
![Screenshot 2020-07-20 at 10 29 58](https://user-images.githubusercontent.com/5824755/88476106-51789380-cf3e-11ea-9295-130c6dbbd3ac.png)

**Comment by jerryyin** [comment]
Thanks for reporting this issue @hohteri. Would it be possible for you to test this in an earlier ROCm version, say ROCm 3.3? Full instructions on installation and uninstallation please refer to [here](https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html).

Or a quick way to experiment is to use an existing docker: `docker pull rocm/tensorflow-autobuilds:rocm3.3-latest`

**Comment by hohteri** [comment]
```
2020-07-30 20:10:05.462991: I tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:188] RocmTracerEvent dropped (correlation_id=11869,) : logical device id >= num gpus.
2020-07-30 20:10:05.464823: I tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:953] GpuTracer stopped
'+fp64-fp16-denormals' is not a recognized feature for this target (ignoring feature)
'-fp32-denormals' is not a recognized feature for this target (ignoring feature)
2020-07-30 20:17:59.830120: I tensorflow/core/common_runtime/gpu_fusion_pass.cc:505] ROCm Fusion is enabled.
```

First 30+ runs done without issues, rocm3.3 seems a bit more verbose especially the last line is printed 9x before each epoch. Also training in our case is about 25% faster than with 3.5.x.

**Comment by jerryyin** [comment]
Thanks for reporting back. The tensor dimension issue is a known defect root caused in Hip runtime, we have internal ticket to track it. The workaround as of now is to keep using ROCm3.3, until a newer ROCm version is rolled out and defect fixed.

**Comment by hohteri** [comment]
I spoke too early (or at least it is more stable than with rocm3.5.x), at run 21 (that is 50 epochs per run), it got stuck after successfully training a model and then running a small evaluation batch. It finished the evaluation batch but got stuck. Weird thing here is that with rocm3.5.x if the execution got stuck It was (almost) always in middle of training not our separate evaluation set. Also if checked with `rocm-smi` the GPUs were idling but with rocm3.3, the other GPU is doing something:

`========================ROCm System Management Interface========================
================================================================================
GPU  Temp   AvgPwr  SCLK     MCLK    Fan     Perf  PwrCap  VRAM%  GPU%  
0    54.0c  58.0W   1802Mhz  351Mhz  21.96%  auto  250.0W    3%   100%  
1    33.0c  19.0W   701Mhz   351Mhz  21.96%  auto  250.0W    3%   0%    
================================================================================
==============================End of ROCm SMI Log ==============================`

**Comment by hohteri** [comment]
We can seem some seldom freezes even with ROCm 3.3 especially when running Keras tuner (tf2.2). If there is a private way of sharing code and training data, we are happy to do that.

**Comment by jerryyin** [comment]
@hohteri Please check you email. Thanks.

**Comment by jerryyin** [comment]
We confirmed that the issue does not exist when using ROCm 3.7 with prefetch. Closing the issue now.
`trainData = trainData.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)`
